### ISSUE TITLE
webhook profile support

### ISSUE THREAD

**Issue by GunnyWaffle** [replaceitem/carpet-discarpet#23]
I dig the new webhook feature, it makes a lot of sense! I was hoping to use it in place of my POSTs to webhooks, but unfortunately found it lacking.

Right now, I use Scarpet Additions and the following function to send webhooks to my Discord server:
```
_webhook(player, message) -> (
	body = {
		'content' -> message,
		'embeds' -> null,
		'allowed_mentions' -> { 'parse' -> [] },
	};
	
	if (player != null,
		body:'username' = player;
		body:'avatar_url' = str(global_avatar_api, player~'uuid');
	);
	
	task(_(b)->http('POST', global_webhook, 10000, 10000, b), encode_json(body));
);
```
I tested this code with the new `dc_send_message` `dc_webhook` target, and it worked. Though the avatar and username were unchanged, which tells me that profile fields are not supported right now.

[disohook](https://discohook.org) is what I use to test and develop webhooks. Would it be possible to see full webhook field support, as displayed by discohook's JSON Data Editor?

Thanks again for the continued development of this mod!

**Comment by replaceitem** [comment]
Does [`dc_update_webhook(webhook, ...)`](https://github.com/replaceitem/carpet-discarpet/blob/master/docs/Full.md#dc_update_webhookwebhook-webhookbuilder) not work for that?

**Comment by GunnyWaffle** [comment]
I can see that technically working, though with two concerns due to how I am using this webhook.
The webhook in my app is used to broadcast chat, events, advancements, deaths, announcements, and more. There is a lot of traffic. The base avatar/name represents the bot, and if it is a player event it uses the player (such as chat messages).

With that comes two concerns with `dc_update_webhook`:
1) Is there a potential for race conditions? Where another webhook is sent between the updating of the webhook and a sending of the message. The concern arises because this decouples the webhook profile from that of the message.
2) This could flood the audit log, since I would be calling `dc_update_webhook` every few seconds.

I'm still able to use HTTP POST, so not a deal breaker, just an edge case I find myself in.

**Comment by replaceitem** [comment]
Adding profile changing aspects to the same api call as send_message would make the scarpet syntax quite messy i think. In most cases setting with update_webhook should be good enogh.
You can also wait for the webhook to be updated, since update_webhook is blocking.
Also, why would you want to update the profile every few seconds anyway?
If you have multiple players, it would probably make more sense to have a webhook for each.

**Comment by GunnyWaffle** [comment]
Even though it is blocking, you can encounter the following race condition unless you wrap all webhook posts with a `synchronize` call:
1) player A joins
2) `dc_update_webhook` to show player A joining
3) `dc_send_message` post join event to webhook
4) player B says 'hey'
5) player C says 'hello'
6) `dc_update_webhook` to show player B talking
7) `dc_update_webhook` to show player C talking
8) `dc_send_message` post B's message to webhook (as player C)
9) `dc_send_message` post C's message to webhook (as player C)

Unfortunately Discord limits you to 10 webhooks per channel, otherwise I would be managing a webhook per player to avoid cache issues with avatar changes. The support for per-message profile changes exists for this reason, to avoid using the Modify Webhook endpoint for each message sent.

I can get around the race condition by using `synchronize` in my scarpet app. Though this hits that endpoint a bit harder now and floods our audit log as well.

I guess it's not as simple as the POST fields in my example then, huh? :<

**Comment by replaceitem** [comment]
Oh, i did now know about that 10 webhook limit. In that case it makes sense to add that, i will see how it is implemented in javacord and probably add it 👍

**Comment by replaceitem** [comment]
Fixed in a27bba3dfc9d16c780f4d36cdd35b10e52fc311a
https://github.com/replaceitem/carpet-discarpet/releases/tag/v1.2.1
